### PR TITLE
Ignore 32bit build option when using msvc

### LINKED ISSUE #447: CMakeLists.txt BENCHMARK_BUILD_32_BITS fails with MSVC
CMakeLists.txt has an option BENCHMARK_BUILD_32_BITS. This works when gcc is used, but it fails with MSVC. In fact this option tries to add -m32 to the compiler options, and this is gcc-specific and doesn't work with MSVC.

With MSVC this option doesn't really make sense, because the CMake MSVC generator determines whether the build is 32-bit or 64-bit. An MSVC user probably doesn't want or expect that setting BENCHMARK_BUILD_32_BITS would make the build fail. Therefore, I propose that BENCHMARK_BUILD_32_BITS be ignored when using MSVC. Doing this is a trivial change to CMakeLists.txt, moving the -m32 stuff so that it is not used with MSVC.

Here's a diff:

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4265a6c..74fe2ea 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -37,10 +37,6 @@ include(CheckCXXCompilerFlag)
 include(AddCXXCompilerFlag)
 include(CXXFeatureCheck)

-if (BENCHMARK_BUILD_32_BITS)
-  add_required_cxx_compiler_flag(-m32)
-endif()
-
 if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
   # Turn compiler warnings up to 11
   string(REGEX REPLACE "[-/]W[1-4]" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
@@ -72,6 +68,10 @@ if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
     set(CMAKE_EXE_LINKER_FLAGS_MINSIZEREL "${CMAKE_EXE_LINKER_FLAGS_MINSIZEREL} /LTCG")
   endif()
 else()
+  if (BENCHMARK_BUILD_32_BITS)
+    add_required_cxx_compiler_flag(-m32)
+  endif()
+
   # Try and enable C++11. Don't use C++14 because it doesn't work in some
   # configurations.
   add_cxx_compiler_flag(-std=c++11)

### PR DESCRIPTION
**Pull request by aJetHorn**
Resolves #447

### CODE REVIEW

**Comment by EricWF** [comment]
This seems like an odd approach to me. Why are people specifying this option for compilers that don't support it?

Looking at how LLVM handles this, I think their approach is more correct. For example, IDK if your patch correctly handles clang-cl on Windows. Perhaps the check should be:
```
 if( CMAKE_SIZEOF_VOID_P EQUAL 8 AND NOT WIN32 )
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,7 +48,8 @@ include(CheckCXXCompilerFlag)
 include(AddCXXCompilerFlag)
 include(CXXFeatureCheck)
 
-if (BENCHMARK_BUILD_32_BITS)
+# Exclude -m32 when using MSVC as it is incompatible
+if (BENCHMARK_BUILD_32_BITS AND NOT "${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
```
Comment (chfast) [inline CMakeLists.txt]: You can simplify it to `NOT MSVC`.

**Comment by dominichamon** [comment]
I don't mind the belt-and-braces approach here. Can you simplify the check as per the comment from @chfast?

**Comment by EricBackus** [comment]
I'd really like to see this merged into the official release, as it fixes a build issue that we've run into several times within Keysight. Any chance this can be merged in?

**Comment by Croydon** [comment]
@dominichamon @chfast Is simplifying this condition the only reason why this didn't get merged yet?

If so I would be happy to open a new pull request.

**Comment by dominichamon** [comment]
@Croydon yes.

**Pull request closed without merging**